sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala

```diff
@@ -181,6 +181,46 @@ object HiveAnalysis extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Add a repartition by dynamic partition columns before insert Hive table.
+ *
+ * Note that, this rule must be run after `HiveAnalysis`.
+ */
+object RepartitionBeforeInsertHiveTable extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.repartitionBeforeInsert) {
```

Comment:
I think this should narrow the scope with `adaptiveExecutionEnabled && coalesceShufflePartitionsEnabled`